File: go/vulnerable_app.go
package main

import (
	"crypto/md5"
	"database/sql"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"os/exec"

	_ "github.com/go-sql-driver/mysql"
)

// Hardcoded credentials
const (
	DBPassword      = "SuperSecret123!"
	APIKey          = "sk_live_4eC39HqLyjWDarhtT657tMo5k"
	AWSAccessKey    = "AKIAIOSFODNN7EXAMPLE"
	AWSSecretKey    = "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
	JWTSecret       = "my-jwt-secret-key"
	PrivateKey      = `-----BEGIN RSA PRIVATE KEY-----
MIIEpAIBAAKCAQEA0Z3VS5JJcds3xfn/ygWyF6R3r4Lv7m8EqFLYrTZY...
-----END RSA PRIVATE KEY-----`
)

var db *sql.DB

func init() {
	var err error
	// Hardcoded database credentials
	db, err = sql.Open("mysql", "root:admin123@tcp(localhost:3306)/mydb")
	if err != nil {
		log.Fatal(err)
	}
}

// SQL Injection vulnerability
func getUser(w http.ResponseWriter, r *http.Request) {
	userID := r.URL.Query().Get("id")
	// Vulnerable to SQL injection
	query := fmt.Sprintf("SELECT * FROM users WHERE id = %s", userID)
	rows, err := db.Query(query)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	defer rows.Close()
	fmt.Fprintf(w, "User data retrieved")
}

// Command Injection
func pingHost(w http.ResponseWriter, r *http.Request) {
	host := r.URL.Query().Get("host")
	// Vulnerable to command injection
	cmd := exec.Command("sh", "-c", fmt.Sprintf("ping -c 1 %s", host))
	output, err := cmd.CombinedOutput()
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	w.Write(output)
}

// Path Traversal
func readFile(w http.ResponseWriter, r *http.Request) {
	filename := r.URL.Query().Get("file")
	// Vulnerable to path traversal
	data, err := ioutil.ReadFile("/var/data/" + filename)
	if err != nil {
		http.Error(w, err.Error(), http.StatusNotFound)
		return
	}
	w.Write(data)
}

// Weak cryptography
func hashPassword(password string) string {
	// Using weak MD5
	hash := md5.Sum([]byte(password))
	return fmt.Sprintf("%x", hash)
}

// SSRF vulnerability
func fetchURL(w http.ResponseWriter, r *http.Request) {
	url := r.URL.Query().Get("url")
	// Vulnerable to SSRF
	resp, err := http.Get(url)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	defer resp.Body.Close()
	body, _ := ioutil.ReadAll(resp.Body)
	w.Write(body)
}

// Insecure random number generation
func generateToken() string {
	// Using math/rand instead of crypto/rand
	import "math/rand"
	return fmt.Sprintf("%d", rand.Intn(9999))
}

// SQL Injection in authentication
func login(w http.ResponseWriter, r *http.Request) {
	username := r.FormValue("username")
	password := r.FormValue("password")
	
	// Vulnerable SQL injection
	query := fmt.Sprintf("SELECT * FROM users WHERE username='%s' AND password='%s'",
		username, password)
	
	rows, err := db.Query(query)
	if err != nil {
		http.Error(w, "Login failed", http.StatusUnauthorized)
		return
	}
	defer rows.Close()
	
	if rows.Next() {
		fmt.Fprintf(w, "Login successful")
	} else {
		fmt.Fprintf(w, "Login failed")
	}
}

// Information disclosure
func errorHandler(w http.ResponseWriter, r *http.Request) {
	err := fmt.Errorf("Database error: mysql://root:password@localhost:3306/mydb connection failed")
	// Exposing sensitive error information
	http.Error(w, err.Error(), http.StatusInternalServerError)
	log.Printf("ERROR: %v", err) // Logging sensitive data
}

// Hardcoded API key usage
func callExternalAPI() {
	url := "https://api.example.com/data?key=" + APIKey
	resp, err := http.Get(url)
	if err != nil {
		log.Printf("API call failed: %v", err)
		return
	}
	defer resp.Body.Close()
}

// Insecure cookie settings
func setCookie(w http.ResponseWriter, r *http.Request) {
	cookie := &http.Cookie{
		Name:     "session",
		Value:    "sensitive-data",
		HttpOnly: false, // Accessible via JavaScript
		Secure:   false, // Not HTTPS only
		SameSite: http.SameSiteNoneMode,
	}
	http.SetCookie(w, cookie)
}

// Unvalidated redirect
func redirect(w http.ResponseWriter, r *http.Request) {
	url := r.URL.Query().Get("url")
	// Open redirect vulnerability
	http.Redirect(w, r, url, http.StatusFound)
}

// Logging sensitive data
func processPayment(w http.ResponseWriter, r *http.Request) {
	cardNumber := r.FormValue("card")
	cvv := r.FormValue("cvv")
	// Logging sensitive payment data
	log.Printf("Processing payment: Card %s, CVV %s", cardNumber, cvv)
	fmt.Fprintf(w, "Payment processed")
}

// Unsafe file operations
func uploadFile(w http.ResponseWriter, r *http.Request) {
	file, header, err := r.FormFile("file")
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	defer file.Close()
	
	// No validation on file type or content
	data, _ := ioutil.ReadAll(file)
	// Unsafe file path
	err = ioutil.WriteFile("/uploads/"+header.Filename, data, 0777)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	fmt.Fprintf(w, "File uploaded")
}

// Hardcoded AWS credentials
func uploadToS3() {
	os.Setenv("AWS_ACCESS_KEY_ID", AWSAccessKey)
	os.Setenv("AWS_SECRET_ACCESS_KEY", AWSSecretKey)
	// AWS operations with hardcoded credentials
}

// Insecure TLS configuration
func createInsecureClient() *http.Client {
	import "crypto/tls"
	
	tr := &http.Transport{
		TLSClientConfig: &tls.Config{
			InsecureSkipVerify: true, // Skipping certificate verification
		},
	}
	return &http.Client{Transport: tr}
}

func main() {
	http.HandleFunc("/user", getUser)
	http.HandleFunc("/ping", pingHost)
	http.HandleFunc("/file", readFile)
	http.HandleFunc("/fetch", fetchURL)
	http.HandleFunc("/login", login)
	http.HandleFunc("/error", errorHandler)
	http.HandleFunc("/cookie", setCookie)
	http.HandleFunc("/redirect", redirect)
	http.HandleFunc("/payment", processPayment)
	http.HandleFunc("/upload", uploadFile)
	
	log.Println("Starting server on :8080")
	log.Printf("Database password: %s", DBPassword) // Logging secrets
	log.Printf("API Key: %s", APIKey)
	
	// Binding to all interfaces
	if err := http.ListenAndServe("0.0.0.0:8080", nil); err != nil {
		log.Fatal(err)
	}
}

File: javascript/vulnerable_app.js
/**
 * Insecure Node.js Application - For Security Scanner Testing Only
 * Contains multiple intentional vulnerabilities
 */

const express = require('express');
const mysql = require('mysql');
const { exec } = require('child_process');
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

const app = express();
app.use(express.json());
app.use(express.urlencoded({ extended: true }));

// Hardcoded credentials
const DB_PASSWORD = 'SuperSecret123!';
const API_KEY = 'sk_live_4eC39HqLyjWDarhtT657tMo5k';
const AWS_ACCESS_KEY = 'AKIAIOSFODNN7EXAMPLE';
const AWS_SECRET = 'wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY';
const JWT_SECRET = 'my-super-secret-jwt-key';

// Database connection with hardcoded credentials
const db = mysql.createConnection({
  host: 'localhost',
  user: 'root',
  password: 'admin123',
  database: 'myapp'
});

// SQL Injection vulnerability
app.get('/user/:id', (req, res) => {
  const userId = req.params.id;
  // Vulnerable to SQL injection
  const query = `SELECT * FROM users WHERE id = ${userId}`;
  db.query(query, (err, results) => {
    if (err) throw err;
    res.json(results);
  });
});

// Command Injection
app.get('/ping', (req, res) => {
  const host = req.query.host || 'localhost';
  // Vulnerable to command injection
  exec(`ping -c 1 ${host}`, (error, stdout, stderr) => {
    res.send(stdout);
  });
});

// Path Traversal
app.get('/file', (req, res) => {
  const filename = req.query.name;
  // Vulnerable to path traversal
  const filepath = path.join(__dirname, 'uploads', filename);
  fs.readFile(filepath, 'utf8', (err, data) => {
    if (err) return res.status(404).send('File not found');
    res.send(data);
  });
});

// XSS vulnerability
app.get('/search', (req, res) => {
  const query = req.query.q || '';
  // Vulnerable to XSS
  res.send(`<h1>Search results for: ${query}</h1>`);
});

// Prototype Pollution
app.post('/merge', (req, res) => {
  const target = {};
  const source = req.body;
  // Vulnerable to prototype pollution
  function merge(target, source) {
    for (let key in source) {
      if (typeof source[key] === 'object') {
        target[key] = merge(target[key] || {}, source[key]);
      } else {
        target[key] = source[key];
      }
    }
    return target;
  }
  const result = merge(target, source);
  res.json(result);
});

// Insecure deserialization
app.post('/deserialize', (req, res) => {
  const serialized = req.body.data;
  // Using eval - dangerous!
  const obj = eval(`(${serialized})`);
  res.json(obj);
});

// NoSQL Injection (MongoDB)
app.post('/login', (req, res) => {
  const { username, password } = req.body;
  // Vulnerable to NoSQL injection
  const query = {
    username: username,
    password: password
  };
  // db.collection('users').findOne(query) would be vulnerable
  res.send('Login processed');
});

// Weak cryptography
function hashPassword(password) {
  // Using weak MD5
  return crypto.createHash('md5').update(password).digest('hex');
}

// Insecure random token generation
function generateToken() {
  // Using Math.random() - not cryptographically secure
  return Math.random().toString(36).substring(7);
}

// SSRF vulnerability
app.get('/fetch', async (req, res) => {
  const url = req.query.url;
  const https = require('https');
  // Vulnerable to SSRF
  https.get(url, (response) => {
    let data = '';
    response.on('data', chunk => data += chunk);
    response.on('end', () => res.send(data));
  });
});

// Regular Expression Denial of Service (ReDoS)
app.get('/validate', (req, res) => {
  const input = req.query.input;
  // Catastrophic backtracking
  const regex = /^(a+)+$/;
  const isValid = regex.test(input);
  res.json({ valid: isValid });
});

// Insecure file upload
app.post('/upload', (req, res) => {
  const file = req.files.upload;
  // No validation on file type or size
  file.mv(`./uploads/${file.name}`, (err) => {
    if (err) return res.status(500).send(err);
    res.send('File uploaded');
  });
});

// Mass assignment vulnerability
app.post('/user/update', (req, res) => {
  const userId = req.body.id;
  // Directly using user input - allows mass assignment
  const updates = req.body;
  // Could allow updating isAdmin, role, etc.
  res.json({ message: 'User updated', updates });
});

// Hardcoded encryption key
const ENCRYPTION_KEY = 'this-is-my-encryption-key-1234567890ab';

// Information disclosure in error
app.get('/error', (req, res) => {
  try {
    throw new Error('Database connection failed at mysql://root:password@localhost:3306/mydb');
  } catch (e) {
    // Exposing stack trace and sensitive info
    res.status(500).send(e.stack);
  }
});

// Using vulnerable packages
// lodash 4.17.15 has prototype pollution
const _ = require('lodash');

// Cookie without security flags
app.get('/set-cookie', (req, res) => {
  res.cookie('session', 'sensitive-data', {
    httpOnly: false,  // Accessible via JavaScript
    secure: false,    // Not HTTPS only
    sameSite: 'none'  // No CSRF protection
  });
  res.send('Cookie set');
});

// CORS misconfiguration
app.use((req, res, next) => {
  // Allowing all origins
  res.header('Access-Control-Allow-Origin', '*');
  res.header('Access-Control-Allow-Credentials', 'true');
  next();
});

// Logging sensitive data
app.post('/process-payment', (req, res) => {
  const { cardNumber, cvv, amount } = req.body;
  // Logging sensitive payment data
  console.log(`Processing payment: Card ${cardNumber}, CVV ${cvv}, Amount ${amount}`);
  res.json({ success: true });
});

// Unvalidated redirect
app.get('/redirect', (req, res) => {
  const url = req.query.url;
  // Open redirect vulnerability
  res.redirect(url);
});

// Server-side template injection
app.get('/template', (req, res) => {
  const name = req.query.name;
  const ejs = require('ejs');
  // Vulnerable to template injection
  const template = `<h1>Hello <%= name %></h1>`;
  const html = ejs.render(template, { name });
  res.send(html);
});

// Running with debug enabled
app.listen(3000, '0.0.0.0', () => {
  console.log('Server running on port 3000');
  console.log(`Database password: ${DB_PASSWORD}`); // Logging secrets
  console.log(`API Key: ${API_KEY}`);
});

module.exports = app;
